[TF] Tensor.replacing: unit test and implicit broadcast

artste opened the pull request:
This pull request contains the following additions:
1) Unit test for tensor.replacing(with,where) fix (#24635)
1) Implicit broadcast for `with` parameter, with relative unit test.

Example:
```swift
// You can write this:
let modified = orig.replacing(with: Tensor<Float>(-1), where: orig.<3)

// Instead of this:
let modified_no_bst = orig.replacing(with: Tensor<Float>(-1).broadcast(like: orig), where: orig.<3)
```

artste committed:
[TF] Added unit test to test `Tensor.replacing(with,where)` (#24635)

Added a unit test to test Tensor.replacing functionality.

artste committed:
[TF] Added implicit broadcast to `with` parameter in `Tensor.replacing(with,where)`. Test included.

Added implicit broadcasting to `with` parameter to simplify syntax.
Example:
  `let modified = original.replacing(with: Tensor<Float>(-1), where: original.<3)`

rxwei commented on test/TensorFlowRuntime/tensor.swift:
```diff
@@ -778,6 +778,25 @@ func testShapeGetter4() {
+}
+TensorTests.testAllBackends("Replacing1", testReplacing1)
+
+@inline(never)
```
No need to mark these as `@inline(never)` (some other tests were written this way for historical technical constraints). And could you use the trailing closure syntax instead? 

```swift
TensorTests.testAllBackends("replacing(with:where:) with broadcasting") {
   let original = Tensor<Float>([[1.0, 0.0], [3.0, 0.0], [2.0, 3.0], [1.0, 0.0]])
   let modified = original.replacing(with: Tensor<Float>(-1), where: original.<3)
   let expected = Tensor<Float>([[-1.0, -1.0], [3.0, -1.0], [-1.0, 3.0], [-1.0, -1.0]])
   expectEqual(expected, modified)
}
```

artste replied:
Great, this is way better.

rxwei commented on test/TensorFlowRuntime/tensor.swift:
```diff
@@ -778,6 +778,25 @@ func testShapeGetter4() {
+
+@inline(never)
+func testReplacing2() {
+  // Implicit broadcast
```
This comment can be included in the test name, as demonstrated in the comment above.

rxwei commented on test/TensorFlowRuntime/tensor.swift:
```diff
@@ -778,6 +778,25 @@ func testShapeGetter4() {
 }
 TensorTests.testAllBackends("ShapeGetter4", testShapeGetter4)
 
+@inline(never)
```
Similarly, `@inline(never)` should be dropped and we should use the trailing closure syntax.

```swift
TensorTests.testAllBackends("replacing(with:where:)") {
   let original = Tensor<Float>([[1.0, 0.0], [3.0, 0.0], [2.0, 3.0], [1.0, 0.0]])
   let modified = original.replacing(with: Tensor<Float>(-1).broadcast(like: original), where: original.<3)
   let expected = Tensor<Float>([[-1.0, -1.0], [3.0, -1.0], [-1.0, 3.0], [-1.0, -1.0]])
   expectEqual(expected, modified)
}
```

rxwei replied:
Also, could you please make sure these tests fit within 80 columns?

artste committed:
[TF] Improved Tests for `Tensor.replacing(with:where:)`

+ Removed `@inline(never)`.
+ Refactor using trailing closure syntax.
+ All lines in 80 columns.

artste commented:
DONE:
+ Removed `@inline(never)`.
+ Refactor using trailing closure syntax.
+ All lines in 80 columns.

rxwei commented:
@artste We've recently moved all TensorFlow APIs from apple/swift to [tensorflow/swift-apis](github.com/tensorflow/swift-apis). Would you mind moving this PR to that library instead? Sorry for the inconvenience!

dan-zheng commented:
This PR seems inactive so I'll close it now. `TensorFlow` APIs have been moved to [https://github.com/tensorflow/swift-apis](https://github.com/tensorflow/swift-apis).
